back/src/Model/GameRoom.ts

```diff
@@ -318,4 +376,114 @@ export class GameRoom {
     public emitEmoteEvent(user: User, emoteEventMessage: EmoteEventMessage) {
         this.positionNotifier.emitEmoteEvent(user, emoteEventMessage);
     }
+
+    public addRoomListener(socket: RoomSocket) {
+        this.roomListeners.add(socket);
+    }
+
+    public removeRoomListener(socket: RoomSocket) {
+        this.roomListeners.delete(socket);
+    }
+
+    /**
+     * Connects to the admin server to fetch map details.
+     * If there is no admin server, the map details are generated by analysing the map URL (that must be in the form: /_/instance/map_url)
+     */
+    private static async getMapDetails(roomUrl: string): Promise<MapDetailsData> {
+        if (!ADMIN_API_URL) {
+            const roomUrlObj = new URL(roomUrl);
+
+            const match = /\/_\/[^/]+\/(.+)/.exec(roomUrlObj.pathname);
+            if (!match) {
+                console.error("Unexpected room URL", roomUrl);
+                throw new Error('Unexpected room URL "' + roomUrl + '"');
+            }
+
+            const mapUrl = roomUrlObj.protocol + "//" + match[1];
+
+            return {
+                mapUrl,
+                policy_type: 1,
+                textures: [],
+                tags: [],
+            };
+        }
+
+        const result = await adminApi.fetchMapDetails(roomUrl);
+        if (!isMapDetailsData(result)) {
+            console.error("Unexpected room details received from server", result);
+            throw new Error("Unexpected room details received from server");
+        }
+        return result;
+    }
+
+    private mapPromise: Promise<ITiledMap> | undefined;
+
+    /**
+     * Returns a promise to the map file.
+     * @throws LocalUrlError if the map we are trying to load is hosted on a local network
+     * @throws Error
+     */
+    private getMap(): Promise<ITiledMap> {
+        if (!this.mapPromise) {
+            this.mapPromise = mapFetcher.fetchMap(this.mapUrl);
+        }
+
+        return this.mapPromise;
+    }
+
+    private variableManagerPromise: Promise<VariablesManager> | undefined;
+
+    private getVariableManager(): Promise<VariablesManager> {
+        if (!this.variableManagerPromise) {
+            this.variableManagerPromise = new Promise<VariablesManager>((resolve, reject) => {
+                this.getMap()
```

Comment:
You could just return this.getMap().then(...) instead of using the Promise constructor

Reply:
```
this.variableManagerPromise = this.getMap()
    .then((map) => {
        const variablesManager = new VariablesManager(this.roomUrl, map);
        return variablesManager
            .init().then(() => {
                return variablesManager;
            })
    }).catch((e) => {
        if (e instanceof LocalUrlError) {
            // If we are trying to load a local URL, we are probably in test mode.
            // In this case, let's bypass the server-side checks completely.

            // Note: we run this message inside a setTimeout so that the room listeners can have time to connect.
            setTimeout(() => {
                for (const roomListener of this.roomListeners) {
                    emitErrorOnRoomSocket(
                        roomListener,
                        "You are loading a local map. If you use the scripting API in this map, please be aware that server-side checks and variable persistence is disabled."
                    );
                }
            }, 1000);

            const variablesManager = new VariablesManager(this.roomUrl, null);
            return variablesManager
                .init()
                .then((variablesManager) => {
                    return variablesManager;
                })
        } else {
            return e;
        }
    });

Reply:
We could simplify this even more if variablesManager.init() could return variablesManager or by using async syntax

Reply:
You are 100% right :+1: